clang/lib/Sema/SemaType.cpp

```diff
@@ -6308,11 +6308,11 @@ TypeResult Sema::ActOnTypeName(Declarator &D) {
     CheckExtraCXXDefaultArguments(D);
   }
 
-  if (const AutoType *AutoT = T->getAs<AutoType>())
-    CheckConstrainedAuto(
-        AutoT,
-        TInfo->getTypeLoc().getContainedAutoTypeLoc().getConceptNameLoc());
-
+  if (const AutoType *AutoT = T->getAs<AutoType>()) {
+    AutoTypeLoc Loc = TInfo->getTypeLoc().getContainedAutoTypeLoc();
+    CheckConstrainedAuto(AutoT,
+                         Loc ? Loc.getConceptNameLoc() : SourceLocation());
+  }
```

Comment:
These still have the same issue, but I don't understand why we need this check here.

We obviously can't write stuff like
```C++
using X = auto;
```

So I am missing in what case undeduced auto can appear here, except the weird deleted function error recovery issue.

Reply:
Ah, is this purely for the error cases?

Ie in my example above, you will still get the error, but is that for suppressing the deprecated warning in case that auto was constrained and happened to be the only use of that concept?

Reply:
I did this for consistency, and now updated to use the same approach as in the other place.

Reply:
But my question was, what cases are we handling with this check in `ActOnTypeName`?
I thought 'auto' couldn't appear here.

Reply:
Test with constrained auto in trailing return type doesn't pass otherwise:
https://github.com/llvm/llvm-project/blob/2bdcfbe62cb9a08df4b58a17d44be0a3082df053/clang/test/CXX/drs/cwg24xx.cpp#L78-L81